File: contextflow-core/core_api/routes/export.py
"""
Export endpoints

GET /api/v1/export/cfpack - export in .cfpack format
GET /api/v1/export/ledger - export JSONL Ledger (extension)
"""

from __future__ import annotations

import json
import sqlite3
import hashlib
from datetime import datetime, timezone
from fastapi import APIRouter, Depends, Query
from fastapi.responses import JSONResponse, StreamingResponse

from core_api import __version__
from core_api.dependencies import get_db
from core_api.schemas import (
    CfpackResponse,
    CfpackProject,
    CfpackTurn,
    CfpackCommit,
    CfpackFindings,
    CfpackHash,
    CfpackMeta,
    TurnWindow,
    FacetSnapshot,
    PipelineConfig,
    Rings,
    Ring1,
    Ring2,
    Ring3,
    Entity,
    PreferenceKeyword,
    Segment,
    EvidenceRef,
)
from core_api.errors import project_not_found


router = APIRouter()


@router.get("/cfpack")
async def export_cfpack(
    project_id: str = Query(..., description="project ID"),
    db: sqlite3.Connection = Depends(get_db)
):
    """
    Export project in .cfpack format (single file JSON archive)

    .cfpack is standard JSON format containing complete semantic version history.

    Note: Draft export is part of Agentic Layer extension, not included in current MVP Core.
    """
    cursor = db.cursor()

    # Get project information
    project_row = cursor.execute(
        "SELECT project_id, name, created_at FROM projects WHERE project_id = ?",
        (project_id,)
    ).fetchone()

    if not project_row:
        raise project_not_found(project_id)

    # Get all Turns
    turn_rows = cursor.execute(
        """
        SELECT turn_hash, parent_turn_hash, role, content, rings_json, created_at
        FROM turns
        WHERE project_id = ?
        ORDER BY created_at ASC
        """,
        (project_id,)
    ).fetchall()

    turns = []
    all_keywords = {}  # Used to aggregate findings

    for row in turn_rows:
        rings_data = json.loads(row["rings_json"]) if row["rings_json"] else None

        # Build Rings object
        rings = None
        if rings_data:
            ring1_data = rings_data.get("ring1", {})
            ring2_data = rings_data.get("ring2", {})
            ring3_data = rings_data.get("ring3", {})

            rings = Rings(
                ring1=Ring1(
                    keywords=ring1_data.get("keywords", []),
                    entities=[Entity(**e) for e in ring1_data.get("entities", [])],
                    time_anchor=ring1_data.get("time_anchor"),
                    preference_keywords=[
                        PreferenceKeyword(**pk)
                        for pk in ring1_data.get("preference_keywords", [])
                    ]
                ),
                ring2=Ring2(
                    intent_seed=ring2_data.get("intent_seed"),
                    time_window=ring2_data.get("time_window"),
                    preference_soft=ring2_data.get("preference_soft", []),
                    unknown_slot=ring2_data.get("unknown_slot", []),
                    facets=ring2_data.get("facets", [])
                ),
                ring3=Ring3(
                    segments=[
                        Segment(**s) for s in ring3_data.get("segments", [])
                    ]
                )
            )

            # Aggregate keywords for findings
            for kw in ring1_data.get("keywords", []):
                if kw not in all_keywords:
                    all_keywords[kw] = {"count": 0, "polarity": "neutral"}
                all_keywords[kw]["count"] += 1

        turns.append(CfpackTurn(
            turn_hash=row["turn_hash"],
            parent_turn_hash=row["parent_turn_hash"],
            role=row["role"],
            content=row["content"],
            created_at=row["created_at"],
            rings=rings
        ))

    # Get all Commits
    commit_rows = cursor.execute(
        """
        SELECT commit_hash, parents_json, branch, turn_window_json,
               facet_snapshot_json, pipeline_config_json, created_at
        FROM commits
        WHERE project_id = ?
        ORDER BY created_at ASC
        """,
        (project_id,)
    ).fetchall()

    commits = []
    for row in commit_rows:
        turn_window_data = json.loads(row["turn_window_json"])
        facet_snapshot_data = json.loads(row["facet_snapshot_json"])
        pipeline_config_data = json.loads(row["pipeline_config_json"]) if row["pipeline_config_json"] else None

        # Build facet_snapshot
        facet_snapshots = []
        for fs in facet_snapshot_data:
            evidence = [EvidenceRef(**e) for e in fs.get("evidence", [])]
            facet_snapshots.append(FacetSnapshot(
                facet=fs["facet"],
                text=fs["text"],
                keywords=fs.get("keywords", []),
                evidence=evidence
            ))

        commits.append(CfpackCommit(
            commit_hash=row["commit_hash"],
            parent_hashes=json.loads(row["parents_json"]),
            branch=row["branch"],
            turn_window=TurnWindow(
                start_turn_hash=turn_window_data["start_turn_hash"],
                end_turn_hash=turn_window_data["end_turn_hash"]
            ),
            facet_snapshot=facet_snapshots,
            pipeline_config=PipelineConfig(**pipeline_config_data) if pipeline_config_data else None,
            created_at=row["created_at"]
        ))

    # Build findings (normalized semantic facts)
    aggregated_keywords = [
        {"lemma": kw, "count": data["count"], "polarity": data["polarity"]}
        for kw, data in sorted(all_keywords.items(), key=lambda x: -x[1]["count"])[:50]
    ]

    # Simplified must_have (high-frequency words)
    must_have = [kw for kw, data in all_keywords.items() if data["count"] >= 2][:20]

    findings = CfpackFindings(
        aggregated_keywords=aggregated_keywords,
        must_have=must_have,
        mustnt_have=[],
        evidence_refs=[]
    )

    # Build complete response (build first, calculate hash later)
    cfpack = CfpackResponse(
        version="1.0.0",
        cfpack_schema_version="1.0.0",
        project=CfpackProject(
            project_id=project_row["project_id"],
            name=project_row["name"],
            created_at=project_row["created_at"]
        ),
        turns=turns,
        findings=findings,
        commits=commits,
        hash=None,  # Calculate later
        meta=CfpackMeta(
            exported_at=datetime.now(timezone.utc).isoformat(),
            exported_by=f"core_api@{__version__}"
        )
    )

    # Calculate package hash (based on complete content excluding hash field)
    # Use JCS style: sorted keys, no extra whitespace
    content_for_hash = cfpack.model_dump(exclude={"hash"})
    canonical_json = json.dumps(content_for_hash, sort_keys=True, separators=(',', ':'))
    pack_hash = f"sha256:{hashlib.sha256(canonical_json.encode()).hexdigest()}"

    cfpack.hash = CfpackHash(
        algorithm="sha256-jcs-v1",
        pack_hash=pack_hash
    )

    # Return response with custom MIME type
    return JSONResponse(
        content=cfpack.model_dump(),
        media_type="application/vnd.contextflow.cfpack+json",
        headers={
            "Content-Disposition": f'attachment; filename="{project_id}.cfpack"'
        }
    )


@router.get("/ledger")
async def export_ledger(
    project_id: str = Query(..., description="project ID"),
    db: sqlite3.Connection = Depends(get_db)
):
    """
    Export project in JSONL Ledger format

    One JSON object per line, sorted chronologically, for backup and streaming processing.
    """
    cursor = db.cursor()

    # Check if project exists
    project_row = cursor.execute(
        "SELECT project_id, name, created_at FROM projects WHERE project_id = ?",
        (project_id,)
    ).fetchone()

    if not project_row:
        raise project_not_found(project_id)

    def generate_jsonl():
        # 1. Output project metadata
        yield json.dumps({
            "type": "project",
            "project_id": project_row["project_id"],
            "name": project_row["name"],
            "created_at": project_row["created_at"]
        }) + "\n"

        # 2. Output all conversations
        conversations = cursor.execute(
            """
            SELECT conversation_id, project_id, title, created_at
            FROM conversations
            WHERE project_id = ?
            ORDER BY created_at ASC
            """,
            (project_id,)
        ).fetchall()

        for conv in conversations:
            yield json.dumps({
                "type": "conversation",
                "conversation_id": conv["conversation_id"],
                "project_id": conv["project_id"],
                "title": conv["title"],
                "created_at": conv["created_at"]
            }) + "\n"

        # 3. Output all Turns
        turns = cursor.execute(
            """
            SELECT turn_hash, parent_turn_hash, project_id, conversation_id,
                   role, content, rings_json, created_at
            FROM turns
            WHERE project_id = ?
            ORDER BY created_at ASC
            """,
            (project_id,)
        ).fetchall()

        for turn in turns:
            rings = json.loads(turn["rings_json"]) if turn["rings_json"] else None
            yield json.dumps({
                "type": "turn",
                "turn_hash": turn["turn_hash"],
                "parent_turn_hash": turn["parent_turn_hash"],
                "project_id": turn["project_id"],
                "conversation_id": turn["conversation_id"],
                "role": turn["role"],
                "content": turn["content"],
                "rings": rings,
                "created_at": turn["created_at"]
            }) + "\n"

        # 4. Output all Commits
        commits = cursor.execute(
            """
            SELECT commit_hash, project_id, branch, message, parents_json,
                   turn_window_json, facet_snapshot_json, pipeline_config_json,
                   draft_id, draft_text_hash, created_at
            FROM commits
            WHERE project_id = ?
            ORDER BY created_at ASC
            """,
            (project_id,)
        ).fetchall()

        for commit in commits:
            yield json.dumps({
                "type": "commit",
                "commit_hash": commit["commit_hash"],
                "project_id": commit["project_id"],
                "branch": commit["branch"],
                "message": commit["message"],
                "parent_hashes": json.loads(commit["parents_json"]),
                "turn_window": json.loads(commit["turn_window_json"]),
                "facet_snapshot": json.loads(commit["facet_snapshot_json"]),
                "pipeline_config": json.loads(commit["pipeline_config_json"]) if commit["pipeline_config_json"] else None,
                "draft_id": commit["draft_id"],
                "draft_text_hash": commit["draft_text_hash"],
                "created_at": commit["created_at"]
            }) + "\n"

    return StreamingResponse(
        generate_jsonl(),
        media_type="application/x-ndjson",
        headers={
            "Content-Disposition": f'attachment; filename="{project_id}.jsonl"'
        }
    )

File: contextflow-cli/src/server/routes/export.ts
/**
 * Export API Routes
 *
 * GET /api/v1/export/cfpack - Export project as .cfpack (JSON archive)
 * GET /api/v1/export/ledger - Export project as JSONL ledger
 *
 * Compatible with Python core_api/routes/export.py response format.
 */

import * as crypto from "node:crypto";
import type { ServerResponse } from "node:http";
import type { Router } from "../router";
import { sendJson } from "../router";
import { errorResponse } from "../types";
import { getDb } from "@contextflow/core";

// ============================================================================
// Types (matching Python schemas)
// ============================================================================

interface Entity {
  text: string;
  type: string;
  start?: number;
  end?: number;
}

interface PreferenceKeyword {
  keyword: string;
  polarity: string;
  lemma: string;
}

interface Segment {
  id: string;
  text: string;
}

interface Ring1 {
  keywords: string[];
  entities: Entity[];
  time_anchor?: string;
  preference_keywords: PreferenceKeyword[];
}

interface Ring2 {
  intent_seed?: string;
  time_window?: string;
  preference_soft: string[];
  unknown_slot: string[];
  facets: string[];
}

interface Ring3 {
  segments: Segment[];
}

interface Rings {
  ring1: Ring1;
  ring2: Ring2;
  ring3: Ring3;
}

interface TurnWindow {
  start_turn_hash: string;
  end_turn_hash: string;
}

interface EvidenceRef {
  turn_hash: string;
  segment_id: string;
  similarity_score: number;
}

interface FacetSnapshot {
  facet: string;
  text: string;
  keywords: string[];
  evidence: EvidenceRef[];
}

interface PipelineConfig {
  id: string;
  sha256: string;
}

interface CfpackTurn {
  turn_hash: string;
  parent_turn_hash: string | null;
  role: string;
  content: string;
  created_at: string;
  rings: Rings | null;
}

interface CfpackCommit {
  commit_hash: string;
  parent_hashes: string[];
  branch: string;
  turn_window: TurnWindow;
  facet_snapshot: FacetSnapshot[];
  pipeline_config: PipelineConfig | null;
  created_at: string;
}

interface CfpackProject {
  project_id: string;
  name: string;
  created_at: string;
}

interface CfpackFindings {
  aggregated_keywords: Array<{ lemma: string; count: number; polarity: string }>;
  must_have: string[];
  mustnt_have: string[];
  evidence_refs: EvidenceRef[];
}

interface CfpackMeta {
  exported_at: string;
  exported_by: string;
}

interface CfpackHash {
  algorithm: string;
  pack_hash: string;
}

interface CfpackResponse {
  version: string;
  cfpack_schema_version: string;
  project: CfpackProject;
  turns: CfpackTurn[];
  findings: CfpackFindings;
  commits: CfpackCommit[];
  hash: CfpackHash | null;
  meta: CfpackMeta;
}

// ============================================================================
// Helpers
// ============================================================================

function parseRings(ringsJson: string | null): Rings | null {
  if (!ringsJson) return null;

  try {
    const data = JSON.parse(ringsJson);
    const ring1 = data.ring1 ?? {};
    const ring2 = data.ring2 ?? {};
    const ring3 = data.ring3 ?? {};

    return {
      ring1: {
        keywords: ring1.keywords ?? [],
        entities: (ring1.entities ?? []).map((e: Entity) => ({
          text: e.text,
          type: e.type,
          start: e.start,
          end: e.end,
        })),
        time_anchor: ring1.time_anchor,
        preference_keywords: (ring1.preference_keywords ?? []).map((pk: PreferenceKeyword) => ({
          keyword: pk.keyword,
          polarity: pk.polarity,
          lemma: pk.lemma,
        })),
      },
      ring2: {
        intent_seed: ring2.intent_seed,
        time_window: ring2.time_window,
        preference_soft: ring2.preference_soft ?? [],
        unknown_slot: ring2.unknown_slot ?? [],
        facets: ring2.facets ?? [],
      },
      ring3: {
        segments: (ring3.segments ?? []).map((s: Segment) => ({
          id: s.id,
          text: s.text,
        })),
      },
    };
  } catch {
    return null;
  }
}

function computeJcsHash(obj: unknown): string {
  // JCS (JSON Canonicalization Scheme): sorted keys, no extra whitespace
  const canonical = JSON.stringify(obj, Object.keys(obj as object).sort(), 0);
  const hash = crypto.createHash("sha256").update(canonical).digest("hex");
  return `sha256:${hash}`;
}

// ============================================================================
// Route Registration
// ============================================================================

export function registerExportRoutes(router: Router): void {
  // GET /api/v1/export/cfpack - Export as .cfpack
  router.get("/api/v1/export/cfpack", async (ctx, _req, res) => {
    const projectId = ctx.query.get("project_id");

    if (!projectId) {
      sendJson(res, 400, errorResponse("INVALID_REQUEST", "project_id query param is required"));
      return;
    }

    try {
      const db = getDb();

      // Get project
      const project = db.prepare(
        "SELECT project_id, name, created_at FROM projects WHERE project_id = ?"
      ).get(projectId) as { project_id: string; name: string; created_at: string } | undefined;

      if (!project) {
        sendJson(res, 404, errorResponse("NOT_FOUND", `Project ${projectId} not found`));
        return;
      }

      // Get all turns
      const turnRows = db.prepare(`
        SELECT turn_hash, parent_turn_hash, role, content, rings_json, created_at
        FROM turns_v2
        WHERE project_id = ?
        ORDER BY created_at ASC
      `).all(projectId) as Array<{
        turn_hash: string;
        parent_turn_hash: string | null;
        role: string;
        content: string;
        rings_json: string | null;
        created_at: string;
      }>;

      const turns: CfpackTurn[] = [];
      const allKeywords: Map<string, { count: number; polarity: string }> = new Map();

      for (const row of turnRows) {
        const rings = parseRings(row.rings_json);

        turns.push({
          turn_hash: row.turn_hash,
          parent_turn_hash: row.parent_turn_hash,
          role: row.role,
          content: row.content,
          created_at: row.created_at,
          rings,
        });

        // Aggregate keywords for findings
        if (rings?.ring1.keywords) {
          for (const kw of rings.ring1.keywords) {
            const existing = allKeywords.get(kw);
            if (existing) {
              existing.count++;
            } else {
              allKeywords.set(kw, { count: 1, polarity: "neutral" });
            }
          }
        }
      }

      // Get all commits
      const commitRows = db.prepare(`
        SELECT commit_hash, parents_json, branch, turn_window_json,
               facet_snapshot_json, pipeline_config_json, created_at
        FROM commits_v2
        WHERE project_id = ?
        ORDER BY created_at ASC
      `).all(projectId) as Array<{
        commit_hash: string;
        parents_json: string;
        branch: string;
        turn_window_json: string;
        facet_snapshot_json: string;
        pipeline_config_json: string | null;
        created_at: string;
      }>;

      const commits: CfpackCommit[] = [];
      for (const row of commitRows) {
        const turnWindow = JSON.parse(row.turn_window_json) as TurnWindow;
        const facetSnapshotData = JSON.parse(row.facet_snapshot_json) as FacetSnapshot[];
        const pipelineConfig = row.pipeline_config_json
          ? (JSON.parse(row.pipeline_config_json) as PipelineConfig)
          : null;

        commits.push({
          commit_hash: row.commit_hash,
          parent_hashes: JSON.parse(row.parents_json),
          branch: row.branch,
          turn_window: turnWindow,
          facet_snapshot: facetSnapshotData,
          pipeline_config: pipelineConfig,
          created_at: row.created_at,
        });
      }

      // Build findings
      const aggregatedKeywords = Array.from(allKeywords.entries())
        .map(([lemma, data]) => ({
          lemma,
          count: data.count,
          polarity: data.polarity,
        }))
        .sort((a, b) => b.count - a.count)
        .slice(0, 50);

      const mustHave = Array.from(allKeywords.entries())
        .filter(([, data]) => data.count >= 2)
        .map(([kw]) => kw)
        .slice(0, 20);

      const findings: CfpackFindings = {
        aggregated_keywords: aggregatedKeywords,
        must_have: mustHave,
        mustnt_have: [],
        evidence_refs: [],
      };

      // Build response (without hash first)
      const cfpack: CfpackResponse = {
        version: "1.0.0",
        cfpack_schema_version: "1.0.0",
        project: {
          project_id: project.project_id,
          name: project.name,
          created_at: project.created_at,
        },
        turns,
        findings,
        commits,
        hash: null,
        meta: {
          exported_at: new Date().toISOString(),
          exported_by: "contextflow-cli@1.0.0",
        },
      };

      // Compute hash
      const { hash: _, ...contentForHash } = cfpack;
      const packHash = computeJcsHash(contentForHash);

      cfpack.hash = {
        algorithm: "sha256-jcs-v1",
        pack_hash: packHash,
      };

      // Send response with custom MIME type
      res.writeHead(200, {
        "Content-Type": "application/vnd.contextflow.cfpack+json",
        "Content-Disposition": `attachment; filename="${projectId}.cfpack"`,
        "Access-Control-Allow-Origin": "*",
      });
      res.end(JSON.stringify(cfpack));
    } catch (err) {
      const message = err instanceof Error ? err.message : "Unknown error";
      sendJson(res, 500, errorResponse("EXPORT_FAILED", message));
    }
  });

  // GET /api/v1/export/ledger - Export as JSONL
  router.get("/api/v1/export/ledger", async (ctx, _req, res) => {
    const projectId = ctx.query.get("project_id");

    if (!projectId) {
      sendJson(res, 400, errorResponse("INVALID_REQUEST", "project_id query param is required"));
      return;
    }

    try {
      const db = getDb();

      // Check project exists
      const project = db.prepare(
        "SELECT project_id, name, created_at FROM projects WHERE project_id = ?"
      ).get(projectId) as { project_id: string; name: string; created_at: string } | undefined;

      if (!project) {
        sendJson(res, 404, errorResponse("NOT_FOUND", `Project ${projectId} not found`));
        return;
      }

      // Set streaming headers
      res.writeHead(200, {
        "Content-Type": "application/x-ndjson",
        "Content-Disposition": `attachment; filename="${projectId}.jsonl"`,
        "Access-Control-Allow-Origin": "*",
      });

      // Stream JSONL

      // 1. Project metadata
      res.write(JSON.stringify({
        type: "project",
        project_id: project.project_id,
        name: project.name,
        created_at: project.created_at,
      }) + "\n");

      // 2. Conversations
      const conversations = db.prepare(`
        SELECT conversation_id, project_id, title, created_at
        FROM conversations
        WHERE project_id = ?
        ORDER BY created_at ASC
      `).all(projectId) as Array<{
        conversation_id: string;
        project_id: string;
        title: string | null;
        created_at: string;
      }>;

      for (const conv of conversations) {
        res.write(JSON.stringify({
          type: "conversation",
          conversation_id: conv.conversation_id,
          project_id: conv.project_id,
          title: conv.title,
          created_at: conv.created_at,
        }) + "\n");
      }

      // 3. Turns
      const turns = db.prepare(`
        SELECT turn_hash, parent_turn_hash, project_id, conversation_id,
               role, content, rings_json, created_at
        FROM turns_v2
        WHERE project_id = ?
        ORDER BY created_at ASC
      `).all(projectId) as Array<{
        turn_hash: string;
        parent_turn_hash: string | null;
        project_id: string;
        conversation_id: string;
        role: string;
        content: string;
        rings_json: string | null;
        created_at: string;
      }>;

      for (const turn of turns) {
        const rings = turn.rings_json ? JSON.parse(turn.rings_json) : null;
        res.write(JSON.stringify({
          type: "turn",
          turn_hash: turn.turn_hash,
          parent_turn_hash: turn.parent_turn_hash,
          project_id: turn.project_id,
          conversation_id: turn.conversation_id,
          role: turn.role,
          content: turn.content,
          rings,
          created_at: turn.created_at,
        }) + "\n");
      }

      // 4. Commits
      const commits = db.prepare(`
        SELECT commit_hash, project_id, branch, message, parents_json,
               turn_window_json, facet_snapshot_json, pipeline_config_json,
               draft_id, draft_text_hash, created_at
        FROM commits_v2
        WHERE project_id = ?
        ORDER BY created_at ASC
      `).all(projectId) as Array<{
        commit_hash: string;
        project_id: string;
        branch: string;
        message: string | null;
        parents_json: string;
        turn_window_json: string;
        facet_snapshot_json: string;
        pipeline_config_json: string | null;
        draft_id: string | null;
        draft_text_hash: string | null;
        created_at: string;
      }>;

      for (const commit of commits) {
        res.write(JSON.stringify({
          type: "commit",
          commit_hash: commit.commit_hash,
          project_id: commit.project_id,
          branch: commit.branch,
          message: commit.message,
          parent_hashes: JSON.parse(commit.parents_json),
          turn_window: JSON.parse(commit.turn_window_json),
          facet_snapshot: JSON.parse(commit.facet_snapshot_json),
          pipeline_config: commit.pipeline_config_json
            ? JSON.parse(commit.pipeline_config_json)
            : null,
          draft_id: commit.draft_id,
          draft_text_hash: commit.draft_text_hash,
          created_at: commit.created_at,
        }) + "\n");
      }

      res.end();
    } catch (err) {
      const message = err instanceof Error ? err.message : "Unknown error";
      // If headers already sent, just end
      if (res.headersSent) {
        res.end();
      } else {
        sendJson(res, 500, errorResponse("EXPORT_FAILED", message));
      }
    }
  });
}
